accounts/abi/bind/backend.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package bind
+
+import (
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// ContractCaller defines the methods needed to allow operating with contract on a read
+// only basis.
+type ContractCaller interface {
+	// ContractCall executes an Ethereum contract call with the specified data as
+	// the input. The pending flag requests execution against the pending block, not
+	// the stable head of the chain.
+	ContractCall(contract common.Address, data []byte, pending bool) ([]byte, error)
+}
+
+// ContractTransactor defines the methods needed to allow operating with contract
+// on a write only basis. Beside the transacting method, the remainder are helpers
+// used when the user does not provide some needed values, but rather leaves it up
+// to the transactor to decide.
+type ContractTransactor interface {
+	// Nonce retrieves the current pending nonce associated with an account.
+	AccountNonce(account common.Address) (uint64, error)
+
+	// GasPrice retrieves the currently suggested gas price to allow a timely execution
+	// of a transaction.
+	GasPrice() (*big.Int, error)
+
+	// GasLimit tries to estimate the gas needed to execute a specific transaction.
```

Comment:
Well, implementation wise the RPC backend delegates it to `eth_estimateGas`, the simulated backend executes the transaction on a temp environment directly. I don't think it's really necessary to define how to calculate it (although there's arguable a single sensible way to do it). The only important thing is for the method to return the amount of gas it needs (in the current pending state) to execute the transaction.